osu.Game/Screens/Play/HUD/KPSCounter/KeysPerSecondCounter.cs

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Localisation;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Rulesets.UI;
+using osu.Game.Skinning;
+using osuTK;
+
+namespace osu.Game.Screens.Play.HUD.KPSCounter
+{
+    public class KeysPerSecondCounter : RollingCounter<int>, ISkinnableDrawable
+    {
+        private const float alpha_when_invalid = 0.3f;
+
+        private readonly Bindable<bool> valid = new Bindable<bool>();
+
+        private static readonly KeysPerSecondCalculator calculator = new KeysPerSecondCalculator();
+
+        [Resolved]
+        private GameplayClock? gameplayClock
+        {
+            get => calculator.GameplayClock;
+            set => calculator.GameplayClock = value;
+        }
+
+        [Resolved(canBeNull: true)]
+        private DrawableRuleset? drawableRuleset
+        {
+            get => calculator.DrawableRuleset;
+            set => calculator.DrawableRuleset = value;
+        }
```

Comment:
There's nowhere else in the code that we do something like this. If unsure, look for other examples first before approaching a problem with a new solution. We prefer consistency.